### ISSUE TITLE
Deploy `squared` Airflow DAGs to MWAA

### ISSUE THREAD

**Issue by aaronsteers** [meltano/squared#326]
The K8 deployment is having some issues and it's not a great situation to just have one engineer able to support/troubleshoot kubernetes. We need to create a quick deployment guide for MWAA anyway, so this seems like a good opportunity to deploy the squared repo over to the MWAA managed Airflow service and at the same time prep a guide for our users who may want to do the same.

Note:

- Out of scope would be any change to Metabase, Superset, or other services. This would just be for the scheduled workflows.

**Comment by pnadolny13** [comment]
Whats the main goal of this work? Is it to get a deployment guide and example implementation for MWAA or to fix the issue with Squared.

If its to fix Squared based on what I logged in https://github.com/meltano/squared/issues/325 I dont think this is the right solution:
- The container task runner we're using right now aka KubernetesPodOperator is having issue starting up containers. This might be an easy fix that Ken could resolve in a quick pairing session. I'm sure I can figure it out myself with enough time but I was requesting help since I didnt want to spin my wheels when we have team members with kubernetes experience.
- I havent used MWAA but from a quick skim it looks like it manages our "always on" Airflow components. The default Airflow helm chart we use in Squared hasnt given us any problems. So I dont think this is solving our problem. We need to fix the actual container runner which (it seems) is out of MWAA's scope.
- We might continue to use our cluster as the container runner, unless we want to switch to ECS. ECS has its own management challenges though.
- We would still need to maintain our kubernetes cluster for other always on services, unless we want to move everything out.

This is a side note because I'm curious but I also assumed (maybe incorrectly) that our managed service would use kubernetes, if thats the case, wouldnt we want to build up more internal skills around managing meltano via kubernetes?

**Comment by aaronsteers** [comment]
> Whats the main goal of this work? Is it to get a deployment guide and example implementation for MWAA or to fix the issue with Squared.

I think potentially "both" but you call out a good point regarding the task run operator and/or the cluster being the problem rather than the Airflow deployment itself. We can definitely consider a switch to ECS - and point taken that this could have its own issues.

> This is a side note because I'm curious but I also assumed (maybe incorrectly) that our managed service would use kubernetes, if thats the case, wouldnt we want to build up more internal skills around managing meltano via kubernetes?

It's possible we will use Kubernetes but this is not decided as of yet and our users should not take a dependency on one expectation or the other. Our users should not expect Kubernetes access, for instance, and ideally the UX would ideally not advertise itself one way or another.

What kind of 'Operator' to use from Airflow does become an issue though (K8, ECS, or custom Meltano Remote operator, for instance), since today users have to choose one or the other.

**Comment by aaronsteers** [comment]
@cjohnhanson, @kgpayne, @pnadolny13 - To Pat's point above, does it make sense for us to build and MWAA deployment guide without already having a custom `MeltanoOperator` in Airflow that can target EKS and/or ECS?

Asking another way: if we did have an MWAA deployment guide, what operator type(s) would we suggest?

**Comment by pnadolny13** [comment]
>It's possible we will use Kubernetes but this is not decided as of yet and our users should not take a dependency on one expectation or the other because the backed could change between iterations. Our users should not expect Kubernetes access, for instance, and ideally the UX would ideally not advertise itself one way or another.

@aaronsteers yeah totally agree, to the user they should have no idea and not care what our backend is.

>To Pat's point above, does it make sense for us to build and MWAA deployment guide without already having a custom MeltanoOperator in Airflow that can target EKS and/or ECS?

What are we hoping the `MeltanoOperator` would add? I'm not totally sure but we really just need to call a service that can start a container and pass it the CLI command. There might be something I'm missing but  https://github.com/meltano/squared/blob/main/data/orchestrate/plugins/meltano_k8_operator.py doesnt do a ton so its possible we can get away just documenting how to use the default operators to call Meltano CLI. KubernetestPodOperator/ECSOperator/BashOperator. Same with ECS, we can point them to https://docs.aws.amazon.com/mwaa/latest/userguide/samples-ecs-operator.html to turn on an ECS cluster and give them a code snippet for calling Meltano CLI using the ECSOperator.

I might be missing something and over simplifying? 😄

**Comment by kgpayne** [comment]
@aaronsteers @pnadolny13 IMHO MWAA would be adequate for our use case, but it's a relatively heavy lift for marginal gain. Given that we already run k8s for Superset and the Meltano UI, we are just adding a second concern in MWAA (new deployment process & credentials, new handling of secrets, new project integration etc.) without taking one away. I expect this would increase our maintenance burden, rather than reduce it, especially as MWAA configuration is different to that available for vanilla (local or hosted) Airflow (i.e. its not a lift-and-shift migration).

As to the opportunity to evaluate and document MWAA, I think that makes complete sense for the self-hosting-managed audience who want Airflow, though there may be lower hanging fruit in the GCP Cloud Run examples that @pandemicsyn put together for the more general case. We chose k8s for a different audience - the self-hosting-self-managing crowd (i.e. people who care about top-to-bottom Open Source, with the ability to run locally or switch provider that that offers).

As @pnadolny13 points out, MWAA offers no task-level control over resources, using only the celery task queue depth for autoscaling. This will likely mean us and our users will hit resource constraints using MWAA workers with the `BashOperator` and have no recourse to address them except to bump up worker instance sizes according to the largest task size. To solve this, we can just recommend the existing k8s or ECS operators for use with MWAA, as we do for Squared and as @pnadolny13 suggests, which might suit the self-hosted-managed user at the cost of adding a second service to their architecture.

For Squared though, the lift on that is still relatively heavy given that the alternative fix for our resource constraint issues is to enable autoscaling for our existing k8s cluster. Right now our cluster capacity is statically provisioned, which is why some deployments and tasks fail to launch due to lack of space on the cluster. We had  [#288](https://github.com/meltano/squared/issues/288) to track that, although in looks like @tayloramurphy closed it 🤔

**Comment by aaronsteers** [comment]
@kgpayne, @cjohnhanson, @pnadolny13, @edgarrmondragon - Putting aside the applicability for our own needs for a sec, do we think we could create a compelling deployment reference for our users using MWAA and ECS operators? And do so in a way that would be somewhere between a `4` and `8` for us in terms of time investment?

If not, I'm inclined to go ahead and close this - although doing so seems to be something of a step away from Airflow as our primary recommended deployment option, as I'm not sure what a simpler Airflow deployment reference would look like.

**Comment by kgpayne** [comment]
@aaronsteers implementing a greenfield example/demo project hosted with MWAA and ECS would be around an `8` in my estimation. Migrating what we already have in Squared without disruption adds both complexity and blast radius (especially in the underlying terraform modules linked to from the k8s blog). If you agree that we don't actually have to adopt MWAA/ECS ourselves if the aim is to generate useful docs, that might be a route to go down.

**Comment by tayloramurphy** [comment]
> implementing a greenfield example/demo project hosted with MWAA and ECS would be around an 8 in my estimation.

With that narrow scope, it'd probably be worth it to do then. I'm inclined to _not_ move `squared` over and reserve any migration efforts for our Alpha managed offering. 

@aaronsteers let's chat in our prioritization meeting about this!

**Comment by pnadolny13** [comment]
@aaronsteers @tayloramurphy based on this discussion, do you think this should be closed and/or migrated to a meltano docs issue?
